### PR TITLE
[TEP-0104] Support Task-level Resource Requirements for TaskRun: Part #1 Fields Addition & Validation w/ Docs Updates 

### PR DESCRIPTION
**Pull request by austinzhao-go**
# Changes
/kind feature

The related TEP: 
[TEP-0104: Support Task-level resource limits](https://github.com/tektoncd/community/pull/703/files?short_path=bd66ac0#diff-bd66ac0b7cdb4930cee261f3a042ee9b7dae6c68d2514de56890037176388dab)
[TEP-0104: Scope Updates to Runtime](https://github.com/tektoncd/community/pull/722)
[TEP-0104: Rename "resources" field as "computeResources"](https://github.com/tektoncd/community/pull/726)

The related FR thread: 
[FR: Task-level (and maybe Pipeline-level) resource requests and limits](https://github.com/tektoncd/pipeline/issues/4470)

This PR will include changes as:
- Add the required "computeResources" field for TaskRun
- Add the related validations
- Update the user guide for this new feature 

A follow-up PR will include changes to update compute resources configuration by the specified task-level resource requirements:
- [ref-to-be-added]()

# TODO
- [x] Update this PR only with field addition and webhook validation
- [x] Add alpha FF to guard changes over webhook
- [x] Update related docs
- [x] Confirm approval and merge on [TEP-0104](https://github.com/tektoncd/community/pull/703/files?short_path=bd66ac0#diff-bd66ac0b7cdb4930cee261f3a042ee9b7dae6c68d2514de56890037176388dab), then start merging implementations 

After the scope update & field name update:
- [x] Update field validation  
- [x] Update the related user guidance 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes
_As changes will be delivered in 2/3 PRs, the release notes will be added in [ref-to-be-added]()._

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/resource_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/resource_types.go:73]: We'll want to document this feature really well, because it's not intuitive how these resource requirements will be applied. I'd suggest adding a bit of detail here, and a lot of detail in [docs/compute-resources.md](https://github.com/tektoncd/pipeline/blob/main/docs/compute-resources.md).

File: pkg/apis/pipeline/v1beta1/resource_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/resource_types.go:76]: Also, as detailed in the TEP, these fields should be added to TaskRun and PipelineRun.TaskRunSpecs too.

File: pkg/reconciler/taskrun/taskrun.go
Comment (lbernick) [inline pkg/reconciler/taskrun/taskrun.go:340]: We should apply validation in the webhook as well as the reconciler (the reason is to return errors as early as possible -- for example an invalid task should be rejected on creation, not when it's run in a taskrun). I'd suggest:

- in [task_validation.go](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/task_validation.go), ensure that the task does not have both step-level and task-level resource requirements
- in [taskrun_validation.go](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_validation.go), make the same check as for tasks
- Here in the reconciler code (probably as a helper function in [validate_resources.go](https://github.com/tektoncd/pipeline/blob/c938ee5667fb64ddcf6e6c26e7732b68dde82104/pkg/reconciler/taskrun/validate_resources.go)), validate that the taskrun resource requirements are compatible with the task resource requirements.

**Review by lbernick — COMMENTED** [review]
Thanks Austin! I would suggest just adding the fields, docs, and webhook validation in the initial PR, and updating the reconciler in a subsequent PR.

/hold

until https://github.com/tektoncd/community/pull/703 is accepted

**Comment by austinzhao-go** [comment]
> 

agree. thinking of a bit the complexity, was planning to complete the logic `Task-level` first, then for "runtime overwriting" from `TaskRunSpec`. but let me do this perhaps (for an easier review):
- field addition and validation on current PR
- `Task-level` update on 2nd PR 
- "runtime overwriting" on 3rd PR

File: pkg/apis/pipeline/v1beta1/resource_types.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/resource_types.go:76]: yap. planned to deliver as a follow-up PR (focusing on "runtime overwriting" part)

File: pkg/reconciler/taskrun/taskrun.go
Comment (austinzhao-go) [reply pkg/reconciler/taskrun/taskrun.go:340]: thanks Lee. let me only keep "validation" logic in the reconciler (removing the resources update part) and add validation in webhook.

**Comment by lbernick** [comment]
ok sounds good! Just one more note, new fields should be guarded behind the alpha feature flag

**Comment by lbernick** [comment]
> got this. So the feature flag checks will be mainly/only needed in reconciler code right? (not changes wekbook code before, let me have another check. perhaps also needed there)

The feature flag checks should be included in the webhook validation, and optionally the reconciler validation, for the same reason as before (fail as fast as possible).

File: pkg/reconciler/taskrun/taskrun.go
Comment (austinzhao-go) [reply pkg/reconciler/taskrun/taskrun.go]: closed by moving changes on reconciler into the following PR

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:77]: Some of this commentary could be reworded to focus less on the implementation and more on the user experience.
For example, where you talk about limits being enforced by the container runtime, or validation being performed by the webhook.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:153]: Suggested wording:

Section Name: Specifying compute resources in TaskRuns

Description: You can specify compute resources at the Step level or at the Task level in TaskRuns, but not both.
To specify compute resources at the Step level, use TaskRuns.StepOverrides. <insert example>
To specify compute resources at the Task level, use TaskRuns.Resources. <insert example>

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:120]: This case and the one below should probably be part of the section on TaskRuns.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:182]: Some of this also gets into implementation details. Since there are multiple possible ways to configure container resource requirements to meet task level resource requirements, I wouldn't describe which way we're choosing in the docs.

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipelinerun_types.go:604]: I'd remove the comment about pods here

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation.go:214]: I think the TEP states that we'll consider different resource types separately; so for example you could specify step level for CPU and task-level for memory. (Open to revisiting this decision.)

Also, why use `.Size > 0` instead of a nil check?

File: pkg/apis/pipeline/v1beta1/resource_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/resource_types.go:64]: Suggested wording:

`TaskResources is overloaded to represent both PipelineResources and compute resources. PipelineResources are deprecated and will be removed, so this field will only represent compute resources in the future.`

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:143]: I think you also need to add TaskRun.resources to the api

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:55]: ya. this works with me.

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:153]: agree this idea.

so even one of the mentioned fields is on `PipelineRun`, still be good for us to mention "in `TaskRun`" right? as the field is under `PipelineRun.TaskRunSpecs`

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:182]: (as a previous comment) let me refine around more under a user doc view. 

thanks this point!

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go:604]: copy here. will keep as 
`// Compute Resources required by the Task/TaskRun`

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation.go:214]: for ` .Size > 0`, an empty/stub resources field was populated out to fail this check when I did the local test on cluster. so using this.. a bit dummy way.  😵

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:143]: you mean `TaskRun.spec.resources` right? 
https://github.com/austinzhao-go/pipeline/blob/tep-0104-per-task-resources-requirements/pkg/apis/pipeline/v1beta1/taskrun_types.go#L44

rechecked our "API changes" section, think I understood wrongly as the referred `TaskSpec` field.

Comment (lbernick) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:143]: yes exactly

File: pkg/apis/pipeline/v1beta1/resource_types.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/resource_types.go:76]: closed - by updating comments + docs

**Comment by austinzhao-go** [comment]
thanks @lbernick. pushed changes for this round of reviews. 
will follow 
- TEP PR got merged asap (reached out community slack channel, and will check in the following API WG session)
- further refining feedback on the user guide docs

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md]: closed - updated as `Specifying compute resources in TaskRun/PipelineRun`

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go]: closed - added the field and related validations

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go]: closed - removed words on `Pod`

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md]: closed - cleaned

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:55]: could you please note this feature is in alpha (you need to enable alpha api fields to get it), and that it is not yet functional? Lastly, docs should be added to taskruns.md as well.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:65]: Since this PR just introduces the new field to "TaskRun", let's stick with only documenting that for now. Also, the examples should use "computeResources" rather than "resources".

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:190]: Could you move this example to the section on LimitRanges? If we're going to include an example like the one you have here, it would be good to show both how it works for step-level and how it works for task-level.

File: pkg/apis/pipeline/v1beta1/taskrun_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:80]: suggested docstring: Compute resources to use for this TaskRun

Also, I think this needs to be a pointer to ResourceRequirements, for the same reason as #4913.

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:144]: apis.ErrMultipleOneOf is probably more appropriate here

Also, I'd suggest the following function name + docstring:

// validateTaskRunComputeResources ensures that compute resources are not configured at both the step level and the task level

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:157]: this function is not needed; just check if taskrun.computeResources is a nil pointer

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation_test.go:519]: could you please add a test case that ensures that you can configure both sidecarOverrides and computeResources?

**Comment by lbernick** [comment]
/hold cancel

since https://github.com/tektoncd/community/pull/703 is now implementable

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:55]: got this. let me do an update.

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:65]: yap. (didn't update the doc yet. let me add the field back for PipelineRun and update the doc with our current TEP requirements)

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:190]: good catch. thanks this!

File: pkg/apis/pipeline/v1beta1/taskrun_types.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go:80]: looked into the details and changes. thanks this ref!

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:157]: right. this will apply nicely after updating as a pointer.

File: pkg/apis/pipeline/v1beta1/taskrun_validation_test.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation_test.go:519]: sure! thanks

File: docs/taskruns.md
Comment (lbernick) [inline docs/taskruns.md:331]: I forgot to mention, please add this info to pipelineruns.md as well.

File: docs/compute-resources.md
Comment (lbernick) [reply docs/compute-resources.md:57]: Thanks for marking as alpha! Please also mark that this isn't functional yet.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:105]: For here and below, I would just say "the following TaskRun will be rejected, because it configures both step-level and task-level compute resource requirements".

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:103]: this section is a bit redundant with the info above it.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:363]: I'd prefer not to include examples that state what the resource requirements on containers will end up being, since this is more of an implementation detail.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:323]: after thinking about this a bit more, I think it would be better to just leave out this section for now. It works basically the same way as what's already described in the docs.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:676]: could you please add a test case with sidecar as well?

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:146]: can you use `if len(overrides.Resources) !=0 ...`?

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:149]: the field name here should just be `computeResources`

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/taskrun_validation.go:146]: can you use `if len(overrides.Resources) !=0 ...`?

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:105]: works!

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:363]: was thinking if this could help users understand how their configure will take an effect, so better know how to configure their requirements. how do you think about it?

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:323]: yap. np for this.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:676]: right. missed it, let me add one case for it.

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:146]: hmm seems a bit better. != is a bit more specific condition for >

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:149]: so here mean we leave out the "parent" field, `taskRunSpecs`, for it? 
thinking it perhaps could help users locate the "error" field

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:146]: ya. +1 as the last comment

**Comment by austinzhao-go** [comment]
cc @ywluogg

File: docs/compute-resources.md
Comment (lbernick) [reply docs/compute-resources.md]: hm, on second thought it might be worth including an example-- my main concern is that there are multiple valid ways to implement task-level resource requirements, so I don't necessarily want to tie ourselves to one way in the documentation. but I'll leave this up to you.

**Review by lbernick — APPROVED** [review]

File: pkg/apis/pipeline/v1beta1/taskrun_validation.go
Comment (austinzhao-go) [reply pkg/apis/pipeline/v1beta1/taskrun_validation.go:149]: right! let me get rid of it. thanks Lee.

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md]: this is a good point as the updating logic is not completed now and could be adjusted when going into the details. 

I backed these changes. let me remove them now, and see if I can offer one meaningful example after impled the resource config/update logic.

**Pull request merged**
